seminar-notes/new-retractionsblame.md

```diff
@@ -0,0 +1,657 @@
+
+```
+    A' <:- A
+    B <:+ B'
+    ---
+    A→B <:+ A'→B
+```
+
+The term that computes this is:
+
+```
+    λ(f : A→B) s_B ∘ f ∘ r_A
+```
+
+Christos:
+  isn't this just what Robby and Matthias Blume did in their implementation?
+Max:
+  yes
+M:
+  Christos, you did this too.
+M:
+  in our world (the tech. report version) we use a pair of projections.
+  Does one of them correspond to your sections?
+Max:
+  no it's different
+M:
+  if we can figure that out, we can use topology instead of ...
+Christos:
+  what's the difference?
+Max:
+  This one's typed, you can only get one projection (one idempotent) from this.
+  But there, you have 2 projections. So strictly more information.
+  Maybe this one is insufficent to justify blame.
+M:
+  exactly, you need the 2 for the blame part
+Max:
+  Well you still have 2 functions, so still possible to distinguish
+  But there you really need 2 contracts
+M:
+  maybe we can build more
+Max:
+  maybe you have more junk
+M:
+  In Scott's world, any retract is a type. Just because the type terrorists
+  haven't written them down doesn't mean they won't be useful
+Max:
+  agreed
+
+We get **mostly** the same rules for positive and negative subtyping, but:
+
+```
+    G <:- Dyn
+```
+
+Is NOT true.
+If it was true, that would mean that casting
+
+```
+    <Dyn ⇒ Boolean> (<Boolean ⇒ Dyn> v)
+```
+
+would have to be the identity for any `v`. That's obviously false.
+
+Let's look more at this disagreement. Maybe Wadler and Findler defined blame
+wrong. Maybe this cast should maybe raise negative blame.
+
+Look at the cast from ground types to `Dyn`. It turns out, `G <:- Dyn` was
+implied by the dynamic semantics of casts:
+
+```
+    <G ⇒ Dyn>p v ↦ Dyn_G v
+```
+
+because this rule throws the blame label `p` away.
+So it says no party is ever blamed.
+
+We could change the dynamic semantics to keep `p`.
+
+```
+  <G ⇒ Dyn>p v ↦ Dyn_G(v, p)
+```
+
+And we also need to change the rule that performs a cast:
+
+```
+  <Dyn ⇒ A>p (Dyn_G (v, q)) ↦ <G ⇒ A>p v      --- if G ~ A
+  <Dyn ⇒ A>p (Dyn_G (v, q)) ↦ blame(p, q')    --- otherwise
+```
+
+Before, we always blamed `p`. Reasoning was, `v` is a bad dynamic value.
+Why not blame `q'`, the context, for performing the cast on `v` ?
+
+Christos:
+  Let me see if I understand correctly. If I have a function with a contract
+  that says it expects booleans, then what does this contract mean?
+  Is it:
+  - my callers have an obligation to produce booleans?
+  - I have an obligation to use my argument as a boolean?
+  Is this the distinction you're making?
+Max:
+  yes, I'm saying it's reasonable to blame the function for not using its
+  argument as a boolean
+
+Here's an example, imagine we have two typed printing functions.
+
+```
+    printN : Num → ⊥
+    printB : Bool → ⊥
+```
+
+And a small program that does these casts:
+
+```
+    ⊤ -> B ⇒ Dyn ⇒ N -> ⊥
+```
+
+(BG: maybe,
+
+```
+   ((λ (x : Bool)
+     (printN (<Dyn ⇒ Nat>(<Bool ⇒ Dyn> x))))
+    true)
+```
```

Comment:
To make a client/server distinction. The `x : Bool` specification is important to me.